File: backend/backend/app/schemas/asset.py
# app/schemas/asset.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List

class AssetBase(BaseModel):
    asset_name: str = Field(..., max_length=150)
    asset_type: str = Field(..., max_length=50)
    asset_serial_number: Optional[str] = Field(None, max_length=100)
    asset_model: Optional[str] = Field(None, max_length=100)
    asset_specifications: Optional[str] = None
    assigned_to: Optional[int] = None
    assigned_location: Optional[str] = Field(None, max_length=150)
    status: Optional[str] = Field("available", max_length=20)

class AssetCreate(AssetBase):
    pass

class AssetUpdate(BaseModel):
    asset_name: Optional[str] = Field(None, max_length=150)
    asset_type: Optional[str] = Field(None, max_length=50)
    asset_serial_number: Optional[str] = Field(None, max_length=100)
    asset_model: Optional[str] = Field(None, max_length=100)
    asset_specifications: Optional[str] = None
    assigned_to: Optional[int] = None
    assigned_location: Optional[str] = Field(None, max_length=150)
    status: Optional[str] = Field(None, max_length=20)

class AssetInDBBase(AssetBase):
    asset_id: int
    allocated_on: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None  # Just the ID, no relationship
    updated_by: Optional[int] = None  # Just the ID, no relationship

    class Config:
        from_attributes = True

class Asset(AssetInDBBase):
    user_name: Optional[str] = None
    user_role: Optional[str] = None
    user_email: Optional[str] = None
    user_mobile: Optional[str] = None

class AssetListResponse(BaseModel):
    total: int
    assets: List[Asset]

File: src/admin/types/asset.ts
import { User } from "./user";

// types/asset.ts
export type AssetType = 'laptop' | 'mobile' | 'tablet';
export type UserRole = 'msr' | 'finance' | 'statehead' | 'zonalhead' | 'printer';

export interface Asset {
  asset_id: number;
  asset_name: string;
  asset_type: AssetType;
  asset_serial_number?: string;
  asset_model?: string;
  asset_specifications?: string;
  assigned_to?: number;
  allocated_on?: string;
  status: 'available' | 'allocated' | 'maintenance' | 'retired';
  created_at: string;
  updated_at: string;
  created_by?: number;
  updated_by?: number;
  
  // User details (from backend)
  user_name?: string;
  user_role?: string;
  user_email?: string;
  user_mobile?: string;
}


export interface AssetCreate {
  asset_name: string;
  asset_type: AssetType;
  asset_serial_number?: string;
  asset_model?: string;
  assigned_to?: number;
  assigned_location?: string;
  status?: 'available' | 'allocated';
}

export interface AssetUpdate {
  asset_name?: string;
  asset_type?: AssetType;
  asset_serial_number?: string;
  asset_model?: string;
  assigned_to?: number;
  assigned_location?: string;
  status?: 'available' | 'allocated' | 'maintenance' | 'retired';
}

export interface AssetListResponse {
  total: number;
  assets: Asset[];
}

export interface AssetAllocationForm {
  asset_name: string;
  asset_type: AssetType;
  asset_serial_number: string;
  asset_model: string;
  assigned_to: number;
}


export interface AssetAllocationProps {
  language: "en" | "hi";
}

export interface ToastState {
  message: string;
  type: "success" | "error";
  isVisible: boolean;
}

export interface AssetAllocationModalProps {
  isOpen: boolean;
  onClose: () => void;
  onAssetAllocated?: () => void;
  users: User[];
  language: 'en' | 'hi';
}

export interface AssetTableProps {
  assets: Asset[];
  language: "en" | "hi";
  onUnassign: (assetId: number) => void;
  loading?: boolean;
}

export interface ConfirmationModalProps {
  isOpen: boolean;
  onClose: () => void;
  onConfirm: () => void;
  assetName: string;
  language: "en" | "hi";
}